v2/tools/generator/internal/codegen/pipeline/assemble_oneof.go

```diff
@@ -0,0 +1,198 @@
+
+func AssembleOneOfTypes() *Stage {
+	stage := NewStage(
+		AssembleOneOfTypesID,
+		"Assemble OneOf types from OpenAPI Fragments",
+		func(ctx context.Context, state *State) (*State, error) {
+			assembler := newOneOfAssembler(state.Definitions())
+
+			newDefs := assembler.assembleOneOfs()
+			return state.WithDefinitions(
+					state.Definitions().OverlayWith(newDefs)),
+				nil
+		})
+
+	return stage
+}
+
+type oneOfAssembler struct {
+	// A set of all the TypeDefinitions we're assembling from
+	defs astmodel.TypeDefinitionSet
+	// A map of all the leaf OneOfs, indexed by the root they reference
+	leavesByRoot map[astmodel.TypeName][]astmodel.TypeName
+	// The set of definitions we've modified
+	updatedDefs astmodel.TypeDefinitionSet
+}
+
+// newOneOfAssembler creates a new assembler to build our oneOf types
+func newOneOfAssembler(defs astmodel.TypeDefinitionSet) *oneOfAssembler {
+	result := &oneOfAssembler{
+		defs:        defs,
+		updatedDefs: make(astmodel.TypeDefinitionSet, 100),
+	}
+
+	result.indexLeaves()
+
+	return result
+}
+
+// assembleOneOfs is called to build up the oneOf types
+func (o *oneOfAssembler) assembleOneOfs() astmodel.TypeDefinitionSet {
+	for rootName, leaves := range o.leavesByRoot {
+		err := o.restructureRoot(rootName, leaves)
+		if err != nil {
+			o.saveError(rootName, errors.Wrapf(err, "error restructuring root oneOf %v", rootName))
+			continue
+		}
+
+		err = o.restructureLeaves(rootName, leaves)
+		if err != nil {
+			o.saveError(rootName, errors.Wrapf(err, "error restructuring leaf oneOfs for %v", rootName))
+		}
+	}
+
+	return o.updatedDefs
+}
+
+// restructureRoot embeds names of the leaves into the root and updates our underlying TypeDefinitionSet.
+func (o *oneOfAssembler) restructureRoot(rootName astmodel.TypeName, leaves []astmodel.TypeName) error {
+
+	root, err := o.lookupOneOf(rootName)
+	if err != nil {
+		return errors.Wrapf(err, "root type name %s didn't map to a known OneOf", rootName)
+	}
+
+	// Add all the leaves into the base type so that it knows about them
+	for _, leaf := range leaves {
+		root = root.WithAdditionalType(leaf)
+	}
+
+	// Replace the old root with the new one
+	o.saveOneOf(rootName, root)
+
+	return nil
+}
+
+// restructureLeaves removes the name of the root from each of the leaves and updates our underlying TypeDefinitionSet.
+// We don't need the name reference as they're now embedded in the root.
+// Plus, failing to remove it results in a circular reference.
+func (o *oneOfAssembler) restructureLeaves(root astmodel.TypeName, leaves []astmodel.TypeName) error {
+	for _, leafName := range leaves {
+		leaf, err := o.lookupOneOf(leafName)
+		if err != nil {
+			return errors.Wrapf(err, "leaf type name %s didn't map to a known OneOf", leafName)
+		}
+
+		leaf = leaf.WithoutType(root)
+		o.saveOneOf(leafName, leaf)
+	}
+
+	return nil
+}
+
+// indexLeaves is called once to build up our index of all the oneOf leaf instances
+func (o *oneOfAssembler) indexLeaves() {
```

Comment:
I rewrote the assembler, but took your thoughts on board when I did it.

I wrote the function as a member mostly for scoping, so it doesn't polute the global/package namespace, but this time I used an anonymous receiver and returned the result instead of assigning it directly.